mempool/mev_reap.go

```diff
@@ -0,0 +1,89 @@
+package mempool
+
+import (
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/types"
+)
+
+// CombineSidecarAndMempoolTxs takes the txs reaped from the mempool and sidecar
+// and combines them into one output Txs.
+//
+// param memplTxs - the ReapedTxs resulting from ReapMaxBytesMaxGas on the mempool
+// param sidecarTxs - the ReapedTxs resulting from ReapMaxBytes on the sidecar
+// param maxBytes - max allowed bytes of output txs
+// param maxGas - max allows gas of output txs
+// param logger - a logger
+//
+// returns a types.Txs of the sidecar txs followed by as many mempool txs as possible
+func CombineSidecarAndMempoolTxs(memplTxs, sidecarTxs types.ReapedTxs, maxBytes, maxGas int64, logger log.Logger) types.Txs {
+	var (
+		totalGas    int64
+		runningSize int64
+	)
+
+	txs := make([]types.Tx, 0, (len(memplTxs.Txs) + len(sidecarTxs.Txs)))
+	sidecarTxsMap := make(map[types.TxKey]struct{})
+
+	for i, sidecarTx := range sidecarTxs.Txs {
+		dataSize := types.ComputeProtoSizeForTxs([]types.Tx{sidecarTx})
+
+		// Check total size requirement
+		if maxBytes > -1 && runningSize+dataSize > maxBytes {
+			logger.Info("[mev-tendermint]: Sidecar txs exceeded block maxBytes, falling back to mempool txs")
+			return memplTxs.Txs
```

Comment:
It seems that if a max is reached while iterating through bundles, we return the txs reaped from the mempool. My guess is that this is to guarantee atomicity (at this point the information about bundles is lost), but it is an unoptimal solution IMO. One option could be to check for maxes when reaping bundles.

Reply:
UPDATE: Agreed it is not optimal. Add to report.

Reply:
This exists because the mev-tendermint node expects a sidecar that fits within max gas and bytes. If the sentinel sends a sidecar that exceeds, we let the mev-tendermint node take this as a sign of a malicious or misbehaving sentinel, and thus falls back to its mempool.